### PR TITLE
add tsconfig.json and setup pipeline to use ts compiler

### PR DESCRIPTION
**Pull request by longlho**
- Add tsconfig.json & tsconfig.cjs.json
- Still use `babel` to compile for jest for now
- Adjust import paths to be more correct (no interop)

Part of #1306

### CODE REVIEW

**Comment by leoyli** [comment]
Maybe just use Babel to handle TypeScirpt if babel already being used?! Also, please consider to add `strict`, and `strictNullChecks` flags. :)

**Comment by longlho** [comment]
Yeah I’ll add those flags once we actually move to TS :) babel is supposed
to be gone but I couldn’t get the ts-jest working. Will try to get rid of
it in subsequent migration PRs

**Comment by leoyli** [comment]
Well, I don't see the benefit to use `ts-jest`. I would suggest to separate the concerns, meaning separate their running steps. It's much easier to setup since jest support TS out-of-box since v24.

**Comment by longlho** [comment]
Oh you mean just babel w TS for jest? We wouldn’t get typechecking though
right?

**Comment by leoyli** [comment]
Yes, no type checking during test running so it can be faster to run and quick to fix.

We then can type-check all files (including test files) in a separate process/stage/script.

In such way, we get both from the two world! 💪

**Comment by longlho** [comment]
ok I'm ok w/ that. We can always revisit later :)

**Comment by PerfectPixel** [comment]
I was following this conversation and have some experience. Using babel + jest is fine, BUT you need to make sure your tsconfig and babel are in line. Otherwise your tests will fail if you enable decorators or use an import syntax.
Personally I got around the whole `ts-jest` by using a custom preprocessor. It is super simple, but it also swallows the type-checks. You only need to use inline source-maps to be able to get good coverage reports:

```
const tsc = require('typescript');
const tsConfig = require('./tsconfig.test.json');

module.exports = {
  process(src, path) {
    if (path.endsWith('.ts') || path.endsWith('.tsx')) {
      return tsc.transpile(src, tsConfig.compilerOptions, path, []);
    }
    return src;
  },
};
```

File: package.json [hunk 1]
```diff
@@ -1,155 +1,132 @@
...
+    "ts-jest": "^24.0.2",
+    "tslib": "^1.9.3",
+    "typescript": "^3.5.1",
+    "watchify": "^3.7.0"
+  },
+  "scripts": {
+    "format": "prettier --check *.js scripts/*.js src/{*.js,**/*.js} *.md **/*.md",
+    "format:fix": "prettier --write *.js scripts/*.js src/{*.js,**/*.js} *.md **/*.md",
+    "clean": "rimraf src/en.js coverage/ dist/ lib/ test/renderer.js",
+    "build:lib": "tsc && tsc -p tsconfig.cjs.json",
```
Comment (redonkulus) [inline package.json]: how long does this take running it twice? in development we should add a watcher `dev:watch` or something to make it easier to develop

**Comment by redonkulus** [comment]
Looks like you are getting a bunch of diffs relating to spacing or CLRF vs RF, would be good if that can be cleaned up for future dev's.

File: package.json (hunk 1 above)
Comment (longlho) [reply package.json]: TS compiler already supports incremental build so it's pretty quick as is

**Comment by leoyli** [comment]
@PerfectPixel, yes there are many "gotcha" in TypeScript... especially on finding the "right" flags to start on.  I personally prefer idiomatic JavaScript.

Since `type` can do almost everything as `interface` but more precise and have no overhead on overloading, that makes the only few keywords or language feature I think is worth to using.  `enum` and `decorator` have their good use case, but it is not so idiomatic, they are more on python or Java things. Also `enum` is not supported well in Babel, so a linting rule may required:

```json
    {
        "no-restricted-syntax": [
          "error",
          {
            "selector": "TSEnumDeclaration",
            "message": "Do not use 'enum': it won't be resolved by Babel."
          }
        ]
    }
```

**Comment by DragonRaider5** [comment]
Actually I'm asking myself whether this actually helps new contributors... F.e. TS would scare me away from first time contributions, what do we have tests for? + our APIs are relatively small...

**Pull request merged**